hibernate-community-dialects/src/main/java/org/hibernate/community/dialect/H2LegacyDialect.java

```diff
@@ -0,0 +1,707 @@
+		}
+		if ( getVersion().isSameOrAfter( 1, 4, 198 ) ) {
+			jdbcTypeRegistry.addDescriptorIfAbsent( H2DurationIntervalSecondJdbcType.INSTANCE );
+		}
+	}
+
+	@Override
+	public int getDefaultStatementBatchSize() {
+		return 15;
+	}
+
+	public boolean hasOddDstBehavior() {
+		// H2 1.4.200 has a bug: https://github.com/h2database/h2database/issues/3184
+		return getVersion().isSameOrAfter( 1, 4, 200 );
+	}
+
+	@Override
+	public void initializeFunctionRegistry(QueryEngine queryEngine) {
+		super.initializeFunctionRegistry( queryEngine );
+
+		CommonFunctionFactory functionFactory = new CommonFunctionFactory(queryEngine);
+
+		// H2 needs an actual argument type for aggregates like SUM, AVG, MIN, MAX to determine the result type
+		functionFactory.aggregates( this, SqlAstNodeRenderingMode.NO_PLAIN_PARAMETER );
+		// AVG by default uses the input type, so we possibly need to cast the argument type, hence a special function
+		functionFactory.avg_castingNonDoubleArguments( this, SqlAstNodeRenderingMode.NO_PLAIN_PARAMETER );
+
+		functionFactory.pi();
+		functionFactory.cot();
+		functionFactory.radians();
+		functionFactory.degrees();
+		functionFactory.log10();
+		functionFactory.rand();
+		functionFactory.truncate();
+		functionFactory.soundex();
+		functionFactory.translate();
+		functionFactory.bitand();
+		functionFactory.bitor();
+		functionFactory.bitxor();
+		functionFactory.bitAndOr();
+		functionFactory.yearMonthDay();
+		functionFactory.hourMinuteSecond();
+		functionFactory.dayOfWeekMonthYear();
+		functionFactory.weekQuarter();
+		functionFactory.daynameMonthname();
+		if ( useLocalTime ) {
+			functionFactory.localtimeLocaltimestamp();
+		}
+		functionFactory.bitLength();
+		functionFactory.octetLength();
+		functionFactory.ascii();
+		functionFactory.octetLength();
+		functionFactory.space();
+		functionFactory.repeat();
+		functionFactory.chr_char();
+		functionFactory.instr();
+		functionFactory.substr();
+		//also natively supports ANSI-style substring()
+		functionFactory.position();
+		functionFactory.trim1();
+		functionFactory.concat_pipeOperator();
+		functionFactory.nowCurdateCurtime();
+		functionFactory.sysdate();
+		functionFactory.insert();
+//		functionFactory.everyAny(); //this would work too
+		functionFactory.everyAny_boolAndOr();
+		functionFactory.median();
+		functionFactory.stddevPopSamp();
+		functionFactory.varPopSamp();
+		if ( getVersion().isSame( 1, 4, 200 ) ) {
+			// See https://github.com/h2database/h2database/issues/2518
+			functionFactory.format_toChar();
+		}
+		else {
+			functionFactory.format_formatdatetime();
+		}
+		functionFactory.rownum();
+		if ( getVersion().isSameOrAfter( 1, 4, 200 ) ) {
+			functionFactory.windowFunctions();
+			functionFactory.listagg( null );
+			if ( getVersion().isSameOrAfter( 2 ) ) {
+				functionFactory.inverseDistributionOrderedSetAggregates();
+				functionFactory.hypotheticalOrderedSetAggregates();
+			}
```

Comment:
Comparing this with the current state of `H2Dialect`, it looks like this should rather be the following:

```suggestion
			if ( getVersion().isSameOrAfter( 2 ) ) {
				functionFactory.listagg( null );
				functionFactory.inverseDistributionOrderedSetAggregates();
				functionFactory.hypotheticalOrderedSetAggregates();
			}
			else {
				// Use group_concat until 2.x as listagg was buggy
				functionFactory.listagg_groupConcat();
			}
```